### PR TITLE
Rename "Stream" to "Buffer", and make "Stream" a pass-through for now

### PR DESCRIPTION
**Pull request by archisgore**
This is designed to resolve this issue:
https://github.com/vulcand/oxy/issues/9

But more so, there've been some occasions where I'd like to use vulcan on long-lived HTTP connections and rather than have the client re-establish connections, I want to pass down data periodically over the same connection.

This change DOES break any kind of sanity checking on request/response sizes, but I do intend to add those back for the stream implementation (since I too want to be able to set those.)

For now, I'm going to deploy my branch onto a vulcand build and see if it even works. Will post results when I have them.

### CODE REVIEW

File: buffer/buffer.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
+/*
+package stream provides http.Handler middleware that solves several problems when dealing with http requests:
```
Comment (klizhentas) [inline buffer/buffer.go]: package buffer

File: buffer/buffer.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+  // sample HTTP handler
+  handler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+    w.Write([]byte("hello"))
+  })
+
+  // Stream will read the body in buffer before passing the request to the handler
+  // calculate total size of the request and transform it from chunked encoding
+  // before passing to the server
+  stream.New(handler)
```
Comment (klizhentas) [inline buffer/buffer.go]: can you change the docs too

Comment (archisgore) [inline buffer/buffer.go]: Yup. Will do ASAP.

File: stream/stream.go [hunk 3]
```diff
@@ -33,37 +33,25 @@ Examples of a streaming middleware:
   stream.New(handler, stream.Retry(`IsNetworkError() && Attempts() <= 2`))
 
 */
-package stream
+package steam
```
Comment (klizhentas) [inline stream/stream.go]: typo?

File: stream/stream_test.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package stream
+package steam
```
Comment (klizhentas) [inline stream/stream_test.go]: typo?

**Comment by archisgore** [comment]
This is tested and working. The "Stream" part of it supports downstream-chunking as well. There is still some bufferring happening somewhere because the "chunks" sent by my test app aren't the same chunks seen by telnet from the front facing side of vulcand.

But this is an IBFS - Initial Basic Feasible Solution. I'll continue to investigate and submit subsequent PRs to make it smoother over the next week. We're already merging this into our middleware to begin pushing it to our beta servers, so this will see a lot of hardening action over the next couple of weeks.

**Comment by nvartolomei** [comment]
> There is still some bufferring happening somewhere because the "chunks" sent by my test app aren't the same chunks seen by telnet from the front facing side of vulcand.

Any ETA for fixing this? I suppose this will allow vulcand to proxy eventsource/server-sent events too.

**Comment by archisgore** [comment]
I added a configurable flush interval which defaults to 100ms for now.

I use the httputil.ReverseProxy to do the actual proxying. Without a flush interval it buffered for quite some time.

I welcome ideas to on how to tell it to flush chunks it gets from the downstream.

**Comment by archisgore** [comment]
I personally think this is safe to merge in. Let me know what I can do to expedite that.

I verified two things:
1. We've been running these changes in production for two weeks now, and haven't seen any problems. We ran heavy movies and large file downloads to test it.
2. I verified that proxies do not have to respect the downstream encoding, and they are allowed to reflow chunks when sending them out upstream.

The best immediately forward-looking solution I can offer right now is, providing an option in vulcand upstream, to allow setting the flush interval period through HttpSettings on the Frontend object in etcd.

The option is already exposed though the forwarder.go in Oxy, so it needs no change here. Once this is merged, we'll move the review to this PR in vulcand: https://github.com/vulcand/vulcand/pull/292

**Comment by klizhentas** [comment]
I will dig into the code, thanks for the PR

**Comment by archisgore** [comment]
Gentle reminder ping. :-) We're running this in production for about six weeks now.

**Comment by archisgore** [comment]
I'm going to merge this in. We tested this for over 2 months in production. I'm happy to revert if something screws up, and also willing to make any code changes should you have suggestions.

**Comment by lunemec** [comment]
The readme is still mentioning usage of Steam (and the usage examples as well).

**Comment by archisgore** [comment]
Fixed. https://github.com/vulcand/oxy/commit/12aa65630360a8ab1827975bd46491a358094cd9

**Pull request merged**